test/conformance/runtime/liveness_probe_test.go

```diff
@@ -0,0 +1,117 @@
+//go:build e2e
+// +build e2e
+
+/*
+Copyright 2022 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtime
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	pkgtest "knative.dev/pkg/test"
+	"knative.dev/pkg/test/spoof"
+	revisionresourcenames "knative.dev/serving/pkg/reconciler/revision/resources/names"
+	v1opts "knative.dev/serving/pkg/testing/v1"
+	"knative.dev/serving/test"
+	"knative.dev/serving/test/conformance/api/shared"
+	v1test "knative.dev/serving/test/v1"
+)
+
+const livenessPath = "/healthz/liveness"
+
+func TestLiveness(t *testing.T) {
+	t.Parallel()
+	if test.ServingFlags.DisableOptionalAPI {
+		t.Skip("Container.livenessProbe is not required by Knative Serving API Specification")
+	}
+	clients := test.Setup(t)
+
+	var testCases = []struct {
+		// name of the test case, which will be inserted in names of routes, configurations, etc.
+		// Use a short name here to avoid hitting the 63-character limit in names
+		// (e.g., "service-to-service-call-svc-cluster-local-uagkdshh-frkml-service" is too long.)
+		name    string
+		handler corev1.Handler
+		sleep   bool
+	}{{
+		name: "httpGet",
+		handler: corev1.Handler{
+			HTTPGet: &corev1.HTTPGetAction{
+				Path: livenessPath,
+			},
+		},
+	}, {
+		name: "httpGetAfterFirstProbe",
+		handler: corev1.Handler{
+			HTTPGet: &corev1.HTTPGetAction{
+				Path: livenessPath,
+			},
+		},
+		sleep: true,
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			names := test.ResourceNames{
+				Service: test.ObjectNameForTest(t),
+				Image:   test.HealthProbes,
+			}
+
+			test.EnsureTearDown(t, clients, &names)
+
+			t.Log("Creating a new Service")
+			resources, err := v1test.CreateServiceReady(t, clients, &names,
+				v1opts.WithLivenessProbe(
+					&corev1.Probe{
+						Handler:          tc.handler,
+						PeriodSeconds:    10,
+						FailureThreshold: 1,
+					}))
+			if err != nil {
+				t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+			}
+
+			// If true sleeping till the first kubelet probe check.
+			if tc.sleep {
+				time.Sleep(15 * time.Second)
```

Comment:
I think this is trying to surface the bug we hit in https://github.com/knative/serving/issues/12462 by making sure the pod stays up for at least 15 seconds even if a liveness probe is configured, but I wouldn't have guessed that without remembering about that issue. I wonder if there's a way to do this that (a) is a bit clearer about what it's testing -- i.e. that a pod with a liveness probe is accessible while that probe returns true (and the probe actually runs), and then is restarted if it fails (b) ideally avoids a hardcoded sleep.

What about if the test image counted how many times its liveness handler had been invoked and returned that number in the response? Then we could `WaitForEndpointState` until the liveness probe had been executed at least N times (avoiding the sleep). We could then - maybe as a follow up - have the test POST to an endpoint on the test image that would cause the liveness check to fail, and assert that the container is properly restarted (this is similar to [how we test readiness probes](https://github.com/knative/serving/blob/main/test/conformance/runtime/readiness_probe_test.go#L258-L265)).

Reply:
makes sense - will do these changes, thanks!

Reply:
@julz just for clarity, if the liveness probe fails and the kubelet does see it, even though it restarts the containers, the readiness probe fails (this is what you documented in your tests I guess), and so the liveness probe check will also hang forever. Is this the beahviour you'd want me to test here? (slightly confused if this is what you meant by asserting that the container is properly restarted.

Reply:
this might be the case where I am just adding bad livenesProbe or readinessProbe and going into the probe pitfalls, but I have tried to capture more of it here: https://github.com/knative/serving/issues/12571